doc/sphinx/proofs/automatic-tactics/auto.rst

```diff
@@ -313,27 +313,56 @@ Creating hint databases
-:cmd:`Create HintDb` have the default setting `Transparent` for both `Variables`
-and `Constants`, while implicitly created databases have the `Opaque` setting.
+by adding a hint to an unknown database.  Databases created with
+:cmd:`Create HintDb` have the default setting `Transparent` (i.e. unfold all)
+for `Constants`, `Projections` and `Variables`, while implicitly created databases
+have the `Opaque` setting (i.e. unfold none).  We recommend using :cmd:`Hint Constants`,
+:cmd:`Hint Projections` and :cmd:`Hint Variables` immediately after :cmd:`Create HintDb`
+to make these settings explicit.
+
+Use :cmd:`Hint Opaque` and :cmd:`Hint Transparent` to control the opacity of
+individual items.  Hint opacity settings influence which hints the search tactics
+try, but they have no effect on how the selected tactic is executed.  The proof
+search tactics use unification to choose which which tactics to try, for example
+whether the goal unifies with a theorem given by :cmd:`Hint Resolve`.  Unfolding
+transparent items is part of the unification process.  The settings are completely
+independent of the non-hint :cmd:`Opaque` and :cmd:`Transparent` settings.
 
 .. cmd:: Create HintDb @ident {? discriminated }
 
-   Creates a new hint database named :n:`@ident`. The database is
-   implemented by a Discrimination Tree (DT) that serves as a filter to select
-   the lemmas that will be applied. When discriminated, the DT uses
-   transparency information to decide if a constant should considered rigid for
-   filtering, making the retrieval more efficient. By contrast, undiscriminated
-   databases treat all constants as transparent, resulting in a larger
-   number of selected lemmas to be applied, and thus putting more pressure on
-   unification.
+   Creates a new hint database named :n:`@ident`.  If the hint database already
+   exists, its current data is dropped
+   (see `#16934 <https://github.com/coq/coq/issues/16934>`_).
+   All constants, variables and projections are set to default to unfoldable (use
+   :cmd:`Hint Constants` to change this).
+
+   By default, hint databases are undiscriminated.  We recommend using `discriminated`.
+   Note: despite the terminology, discrimination trees are always used to pattern match
+   some types of hints.
+
+Performance considerations
+``````````````````````````
+
+The proof search tactics decide which hints to try by first selecting hints
+whose pattern has the same :term:`head constant` as the goal.  Then the selected
+hints are tried as follows (ordered from best performing to worst performing):
+
+- For hints other than :cmd:`Hint Extern`, if the database is discriminated and
+  the pattern doesn't contain unfoldable (transparent) items, the full
+  pattern must match the goal before the hint will be tried.
+- If a hint other than `Hint Extern` contains unfoldable items, it will be tried
+  even if the pattern doesn't match.
```

Comment:
I think this covers the first two bullets:

- For hints other than :cmd:`Hint Extern`, if the database is discriminated,
  all items that are opaque in both the goal and the pattern must match before the
  hint will be tried.

But not sure `discriminated` really matters.  I think the only difference for nondiscriminated is in the details of Hint Constructor, which I don't fully understand.  Is it possible and useful to have types such as `bool2` and `Hint Constructor bool2`?  Or to have arguments in such constructors?

```
Inductive bool2 : Prop :=
  | true2 : bool2
  | false2 : bool2.
Inductive True : Prop := (* for comparison *)
  I : True.

Hint Constructors bool2 : db.
Hint Constructors True : db.
Print HintDb db.

(*
For any goal ->   
For True ->   exact I (cost 0, pattern True, id 0)
For bool2 ->   exact false2 (cost 0, pattern bool2, id 0)
               exact true2 (cost 0, pattern bool2, id 0)
*)

  Goal true2.  (* The term "true2" has type "bool2" which should be Set, Prop or Type. *)
```

Reply:
I like your suggested formulation. Technically, `Hint Extern` is not an exception to the way you describe algorithm. In fact, with Coq 8.21, the patterns of `Hint Extern` are added to the dnet with full opacity. As a consequence, not even the syntactic pattern matching that was visible with `Set Typeclasses Debug` is shown now when the pattern does not match the goal. It would be fine to just remove the mention of `Hint Extern` here.

I am not familiar with `Hint Constructor` at all so I can't speak to that.